miniapps/dpg/pconvection-diffusion.cpp

```diff
@@ -0,0 +1,788 @@
+
+      int dofs = u_fes->GlobalTrueVSize()
+                 + sigma_fes->GlobalTrueVSize()
+                 + hatu_fes->GlobalTrueVSize()
+                 + hatf_fes->GlobalTrueVSize();
+
+      double L2Error = 0.0;
+      double rate_err = 0.0;
+      if (exact_known)
+      {
+         double u_err = u_gf.ComputeL2Error(uex);
+         double sigma_err = sigma_gf.ComputeL2Error(sigmaex);
+         L2Error = sqrt(u_err*u_err + sigma_err*sigma_err);
+         rate_err = (it) ? dim*log(err0/L2Error)/log((double)dof0/dofs) : 0.0;
+         err0 = L2Error;
+      }
+      double rate_res = (it) ? dim*log(res0/gresidual)/log((double)dof0/dofs) : 0.0;
+
+      res0 = gresidual;
+      dof0 = dofs;
+
+      if (myid == 0)
+      {
+         std::ios oldState(nullptr);
+         oldState.copyfmt(std::cout);
+         std::cout << std::right << std::setw(5) << it << " | "
+                   << std::setw(10) <<  dof0 << " | ";
+         if (exact_known)
+         {
+            std::cout << std::setprecision(3) << std::setw(10)
+                      << std::scientific <<  err0 << " | "
+                      << std::setprecision(2)
+                      << std::setw(6) << std::fixed << rate_err << " | " ;
+         }
+         std::cout << std::setprecision(3)
+                   << std::setw(10) << std::scientific <<  res0 << " | "
+                   << std::setprecision(2)
+                   << std::setw(6) << std::fixed << rate_res << " | "
+                   << std::setw(6) << std::fixed << num_iter << " | "
+                   << std::endl;
+         std::cout.copyfmt(oldState);
+      }
+
+      if (visualization)
+      {
+         const char * keys = (it == 0 && dim == 2) ? "cgRjmlk\n" : nullptr;
+         char vishost[] = "localhost";
+         int  visport   = 19916;
+         VisualizeField(u_out,vishost, visport, u_gf,
+                        "Numerical u", 0,0, 500, 500, keys);
+         VisualizeField(sigma_out,vishost, visport, sigma_gf,
+                        "Numerical flux", 501,0,500, 500, keys);
+      }
+
+      if (paraview)
+      {
+         paraview_dc->SetCycle(it);
+         paraview_dc->SetTime((double)it);
+         paraview_dc->Save();
+      }
+
+      if (it == ref)
+      {
+         break;
+      }
+
+      pmesh.GeneralRefinement(elements_to_refine,1,1);
+      for (int i =0; i<trial_fes.Size(); i++)
+      {
+         trial_fes[i]->Update(false);
+      }
+      a->Update();
+
+      coeff_fes->Update();
+      c1_gf.Update();
+      c2_gf.Update();
+      setup_test_norm_coeffs(c1_gf,c2_gf);
+   }
+
+   if (paraview)
+   {
+      delete paraview_dc;
+   }
+
+   delete coeff_fes;
+   delete coeff_fec;
+   delete a;
+   delete tau_fec;
+   delete v_fec;
+   delete hatf_fes;
+   delete hatf_fec;
+   delete hatu_fes;
+   delete hatu_fec;
+   delete sigma_fec;
+   delete sigma_fes;
+   delete u_fec;
+   delete u_fes;
+
+   return 0;
+}
+
+void solution(const Vector & X, double & u, Vector & du, double & d2u)
```

Comment:
There is some duplication of solution and coefficient functions in the serial and parallel versions of each test, which could be eliminated with a separate file for the functions. In some cases, one file may have more general versions or more options, but they could still be combined. I think it would help but would not make a big difference, so I am fine with keeping the current version unless you agree it would simplify the code.

Reply:
Indeed that's the case. I would prefer to keep the files self-contained though